drivers/adc/adc_silabs_iadc.c

```diff
@@ -0,0 +1,688 @@
+	 * (gain + reference combinations + oversampling), so we need to map
+	 * multiple channel configs to these 2 available ADC configs.
+	 */
+	ARRAY_FOR_EACH(data->chan_conf, i) {
+		chan_conf = &data->chan_conf[i];
+
+		if (!chan_conf->initialized || (i != find_lsb_set(channels) - 1)) {
+			continue;
+		}
+
+		res = iadc_find_or_create_adc_config(data, &adc_init_config, chan_conf);
+		if (res < 0) {
+			LOG_DBG("IADC: too many different ADC configurations");
+			return res;
+		}
+
+		chan_conf->iadc_conf_id = res;
+
+		iadc_configure_scan_table_entry(&scan_table.entries[i], chan_conf);
+
+		channels &= ~BIT(i);
+	}
+
+	sl_hal_iadc_init(iadc, &adc_init_config, data->clock_rate);
+	sl_hal_iadc_init_scan(iadc, &scan_init, &scan_table);
+	sl_hal_iadc_set_scan_mask_multiple_entries(iadc, &scan_table);
+
+	return 0;
+}
+
+static int iadc_check_buffer_size(const struct adc_sequence *sequence, uint16_t active_channels,
+				  size_t data_size)
+{
+	size_t needed_buffer_size = active_channels * data_size;
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_DBG("Provided buffer is too small (%u/%u)", sequence->buffer_size,
+			needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+/*
+ * Goal of this function is to have concensius between wanted resolution, oversampling,
+ * the IADC Alignment Table, analog oversampling and digital averaging.
+ *
+ * Formulas:
+ * Output Resolution = 11 + log2(OversamplingRatio × DigitalAveraging)
+ * +------------+-------------+---------------+----------------+-------------+----------------+
+ * | Alignment  | Oversample  | Digital Avg   | Num Samples   | Output Res  | DATA Mask     |
+ * | Setting    | Ratio       |               | Averaged      |             |               |
+ * +------------+-------------+---------------+----------------+-------------+----------------+
+ * | 16-bit     | 2x          | 1x            | 2             | 12 bits     | 0xFFF0        |
+ * | 16-bit     | 8x          | 2x            | 16            | 15 bits     | 0xFFFE        |
+ * | 20-bit     | 2x          | 1x            | 2             | 12 bits     | 0xFFF00       |
+ * | 20-bit     | 16x         | 4x            | 64            | 17 bits     | 0xFFFF8       |
+ * +------------+-------------+---------------+----------------+-------------+----------------+
+ */
+static int iadc_check_oversampling_and_resolution(const struct adc_sequence *sequence,
+						  struct iadc_data *data)
+{
+	int res, ospl;
+
+	const static struct oversampling_table {
+		uint32_t wanted_oversampling;
+		uint8_t analog_oversampling;
+		uint8_t digital_averaging;
+		uint8_t alignment;
+		uint16_t resolution;
+	} ospl_table[] = {
+		{ 0, _IADC_CFG_OSRHS_HISPD2, _IADC_CFG_DIGAVG_AVG1,
+		 _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT12, 12}, /* default */
+		{ 1, _IADC_CFG_OSRHS_HISPD2, _IADC_CFG_DIGAVG_AVG1,
```

Comment:
Nit: `wanted_oversampling` is unused and you can get this information using the table index (rule of the single root of trust). You could make the index explicit with:
```
    .[0] = {
        _IADC_CFG_OSRHS_HISPD2, _IADC_CFG_DIGAVG_AVG1,
        _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT12, 12,
    },
```